File: core/contractsdk/go/example/award_manage/award_manage.go
package main

import (
	"bytes"
	"errors"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/code"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/driver"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/utils"
	utils2 "github.com/xuperchain/xuperchain/core/contractsdk/go/utils"
)

const (
	BALANCEPRE   = "balanceOf_"
	ALLOWANCEPRE = "allowanceOf_"
	MASTERPRE    = "owner"
)

type awardManage struct{}

func (am *awardManage) Initialize(ctx code.Context) code.Response {
	args := struct {
		totalSupply []byte `json:"totalSupply",gt:"0"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}

	caller := ctx.Initiator()
	if len(caller) == 0 {
		return code.Error(errors.New(""))
	}

	if err := ctx.PutObject([]byte(BALANCEPRE+caller), args.totalSupply); err != nil {
		return code.Error(err)
	}

	if err := ctx.PutObject([]byte("totalSupply"), args.totalSupply); err != nil {
		return code.Error(err)
	}

	if err := ctx.PutObject(utils2.Concat(MASTERPRE), utils2.Concat(caller)); err != nil {
		return code.Error(err)
	}

	return code.OK(nil)

}

func (am *awardManage) AddAward(ctx code.Context) code.Response {
	caller := ctx.Initiator()
	master, err := ctx.GetObject([]byte(MASTERPRE))
	if err != nil {
		return code.Error(errors.New(""))

	}
	if !bytes.Equal(master, []byte(caller)) { //TODO equal?
		return code.Error(errors.New(""))
	}
	args := struct {
		Amount []byte `json:"amount"`
		Value  []byte `json:"value"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}

	totalSupply := utils2.Add(args.Amount, args.Value)
	ctx.PutObject([]byte("totalSupply"), totalSupply)
	return code.OK([]byte("ok"))
}

func (am *awardManage) TotalSupply(ctx code.Context) code.Response {

	if value, err := ctx.GetObject([]byte("totalSupply")); err != nil {
		return code.Error(err)
	} else {
		return code.OK(value)
	}

}

func (am *awardManage) Balance(ctx code.Context) code.Response {
	args := struct {
		Caller []byte `json:"caller"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}

	value, err := ctx.GetObject(append([]byte(BALANCEPRE), args.Caller...))
	if err != nil {
		return code.Error(err)
	}
	return code.OK(value)
}

func (am *awardManage) Allowance(ctx code.Context) code.Response {
	args := struct {
		From []byte `json:"from"`
		To   []byte `json:"to"`
	}{}
	utils.Validate(ctx.Args(), &args)

	buf := bytes.NewBufferString(ALLOWANCEPRE)
	buf.Write(args.From)
	buf.WriteString("_")
	buf.Write(args.To)

	value, err := ctx.GetObject(buf.Bytes())
	if err != nil {
		return code.Error(err)
	}
	return code.OK(value)

}

func (am *awardManage) Transfer(ctx code.Context) code.Response {

	args := struct {
		From   []byte `json:"from",required:"true",eq:"",neq:"",lt:"",gt:""`
		To     []byte `json:"to",required:"true"`
		Caller string `json:"caller",required:"true"`
		Token  []byte `json:"token",required:"true"`
		Raw    map[string][]byte
	}{}

	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	if utils2.Compare(args.From, args.Token) == 0 {
		return code.Error(errors.New("can not transfer to yourself"))
	}

	from_balance, err := ctx.GetObject((append([]byte(BALANCEPRE), args.From...)))
	if err != nil {
	}
	if utils2.Compare(from_balance, args.Token) < 0 {
		return code.Error(nil)
	}
	to_balance, err := ctx.GetObject((append([]byte(BALANCEPRE), args.To...)))
	if err != nil {
		return code.Error(err)
	}

	to_balance = utils2.Add(to_balance, args.Token)
	from_balance = utils2.Sub(from_balance, args.Token)
	return code.OK([]byte("ok"))

}

func (am *awardManage) TransferFrom(ctx code.Context) code.Response {
	args := struct {
		From   string `json:"from",required:"true"`
		To     string `json:"to",required:"true"`
		Caller string `json:"caller",required:"true"`
		Token  []byte `json:"token",required:"true"`
	}{}

	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}

	value, err := ctx.GetObject([]byte(ALLOWANCEPRE + args.From + "_" + args.Caller))
	if err != nil {
		return code.Error(err)

	}
	if utils2.Compare(value, args.Token) < 0 {
		return code.Error(utils.ErrBalanceLow)
	}

	from_balance, err := ctx.GetObject([]byte(BALANCEPRE + args.From))
	if err != nil {
		return code.Error(err)
	}
	if utils2.Compare(from_balance, args.Token) < 0 {
		return code.Error(utils.ErrBalanceLow)
	}
	to_balance, err := ctx.GetObject([]byte(BALANCEPRE + args.To))
	if err != nil {
		return code.Error(err)
	}
	_ = to_balance
	allowance_balance := []byte{} //TODO
	allowance_balance = utils2.Sub(allowance_balance, args.Token)
	allowance_key := []byte("") // TODO
	ctx.PutObject(allowance_key, allowance_balance)
	return code.OK(nil)

}

func (am *awardManage) Approve(ctx code.Context) code.Response {
	args := struct {
		From         []byte `json:"from",required:"true"`
		To           []byte `json:"to",required:"true"`
		Caller       []byte `json:"caller",required:"true"`
		Token        []byte `json:"token",required:"true"`
		ValidateFunc func() error
	}{}

	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	value, err := ctx.GetObject(utils2.Concat(ALLOWANCEPRE, args.From, "_", args.To))
	if err != nil {

	}

	from_balance, err := ctx.GetObject(utils2.Concat(BALANCEPRE, args.From))
	if err != nil {
		return code.Error(err)
	}

	if utils2.Compare(value, args.Token) < 0 {
		return code.Error(utils.ErrBalanceLow)
	}
	if utils2.Compare(from_balance, args.Token) < 0 {
		return code.Error(utils.ErrBalanceLow)
	}
	to_balance, err := ctx.GetObject(utils2.Concat(BALANCEPRE, args.To))
	if err != nil {
		return code.Error(err)
	}
	allowance_balance := []byte{} //TODO
	allowance_balance = utils2.Sub(allowance_balance, args.Token)
	allowance_key := []byte("")
	ctx.PutObject(allowance_key, to_balance)
	return code.OK(nil)
}

func main() {
	driver.Serve(new(awardManage))
}

File: core/contractsdk/java/example/award_manage/src/main/java/com/baidu/xuper/example/AwardManage.java
package com.baidu.xuper.example;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

import com.baidu.xuper.Context;
import com.baidu.xuper.Contract;
import com.baidu.xuper.ContractMethod;
import com.baidu.xuper.Driver;
import com.baidu.xuper.Response;
import com.baidu.xuper.utils.ByteUtils;
/**
 * Counter
 */
public class AwardManage implements Contract {
    final String BALANCE = "balanceOf_";
    final String ALLOWANCE = "allowanceOf_";
    final String MASTER = "owner";
    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        String caller = ctx.caller();
        if (caller.length()==0){
            return Response.error("missing caller");
        }
        byte[]totalSupply = ctx.args().get("totalSupply");
        if (totalSupply.length ==0){
            return Response.error("missing totalSupply");
        }
        if (ByteUtils.Less(totalSupply,0) || ByteUtils.Equal(totalSupply,0)){
            return Response.error("totalSupply is overflow");
        }
        ctx.putObject(ByteUtils.Concat(BALANCE,caller),totalSupply);
        ctx.putObject("totalSupply".getBytes(),totalSupply);
        ctx.putObject(MASTER.getBytes(),caller.getBytes());
        return Response.ok("ok".getBytes());
    }
//TODO @fengjin 考虑装饰器模式？？
    private boolean permCheck(Context ctx){
        return true;
    }

    public Response addAward(Context ctx){
        if (!permCheck(ctx)){
            return Response.error("you do not have permission to do this operation")
        }
        byte[] amount = ctx.args().get("amount".getBytes());
        if (!ByteUtils.More(amount,0)){
            return Response.error("amount must be positive");
        }

        byte[] totalSupply = ctx.getObject ("totalSupply".getBytes());
        if (!ByteUtils.More(value,0)){
            return Response.error("totalSupply must positive");
        }
        if (ByteUtils.Less(ByteUtils.Add(totalSupply,amount),0)){
            return Response.error("amount add to total supply overflow");
        }
        ctx.putObject("TotalSupply",BytesUtils.Add(totalSupply,amount));
        byte[] key = ByteUtils.Concat(BALANCE,ctx.caller());
        byte[] value = ByteUtils.Add(ctx.getObject(key),totalSupply)
        ctx.putObject(key,value)); // TODO @fengjin 溢出攻击
        return Response.ok(value);
    }

    public Response totalSupply(Context ctx){
        return Response.ok(ctx.getObject("totalSupply".getBytes()));
    }
    public Response  balance(Context ctx){
        byte[]caller = ctx.getObject("caller".getBytes());
        return Response.ok(ctx.getObject(ByteUtils.Concat(BALANCE,caller)));
    }
    public Response allowance(Context ctx){
        byte[] from = ctx.args().get("from".getBytes());
        byte[] to = ctx.args().get("to".getBytes());
        if (from.length == 0 || to.length == 0){
            return Response.error("missing from or to");
        }
        return ctx.getObject(ByteUtils.Concat(ALLOWANCE.getBytes(),from,"_".getBytes(),to))

    }

    @ContractMethod
    public Response transfer(Context ctx){

        byte[] from = ctx.args().get("from".getBytes());
        byte[] to = ctx.args().get("to".getBytes());
        byte[] token = ctx.args().get("token"); // TODO
        if(from.length==0 || to.length == 0 || token.length ==0){
            return Response.error("missing from or to or token");
        }
        if (ByteUtils.Equal(from,to)){
            return Response.error("you can not transfer to yourself");
        }
        if (ByteUtils.Less(token,0)|| ByteUtils.Equal(token,0)){
            return Response.error("token must be more than 0");
        }


        byte[] from_key = ByteUtils.Concat(BALANCE,from);
        byte[] from_balance = ctx.getObject(from_key);
        if (ByteUtils.Less(from_balance,token)){
            return Response.error("balance not enough");
        }
        byte[] to_key = ByteUtils.Concat(BALANCE,to);
        byte[]to_balance = ctx.getObject(to_key);
        from_balance = Byteutls.Sub(from_balance,token);

        ctx.putObject(from_key,ByteUtils.sub(from_balance,token));
        ctx.putObject(to_key,ByteUtils.add(to_balance,token));
        return Response.ok("transfer success".getBytes());
    }
    @ContractMethod
    public Response transferFrom(Context ctx){

        byte[] from = ctx.args().get("from".getBytes());
        byte[] to = ctx.args().get("to".getBytes());
        byte[] token = ctx.args().get("token"); // TODO
        byte[] caller = ctx.args().get("caller");
        if(from.length==0 || to.length == 0 || token.length ==0 || caller.length==0){
            return Response.error("missing from or to or token");
        }
        if (ByteUtils.Equal(from,to)){
            return Response.error("you can not transfer to yourself");
        }
        if (ByteUtils.Less(token,0)|| ByteUtils.Equal(token,0)){
            return Response.error("token must be more than 0");
        }
        byte[] allowance_key= ByteUtils.concat(ALLOWANCE,from,"_".getBytes(),caller);
        byte[] allowance_value = ctx.getObject(allowance_key);
        if (allowance_value.length == 0){
            return Response.error("you need to add allowance from_to");
        }
        if (ByteUtils.Less(allowance_value,token)) {
            return Response.error("The allowance of from_to not enough");
        }
        byte[] from_key = ByteUtils.concat(BALANCE,from);
        byte[] from_balance = ctx.getObject(from_key);

        if (from_balance.length ==0){
//            TODO
        }

        if (ByteUtils.Less(from_balance,token)){
            return Response.error("the balance of from not enough");
        }
        byte[]to_balance = ctx.getObject(ByteUtils.concat(BALANCE,to)){

        }
        ctx.putObject(from_key,ByteUtils.sub(from_balance,token));
        ctx.putObject(to_key,ByteUtils.add(to_balance,token));
        ctx.putObject(allowance_key,ByteUtils.sub(allowance_value,token));
    }
    @ContractMethod
    public Response approve(Context ctx){
        byte[] from = ctx.args().get("from".getBytes());
        byte[] to = ctx.args().get("to".getBytes());
        byte[] token = ctx.args().get("token"); // TODO
        if(from.length==0 || to.length == 0 || token.length ==0){
            return Response.error("missing from or to or token");
        }
        if (ByteUtils.Equal(from,to)){
            return Response.error("you can not transfer to yourself");
        }
        byte[] from_key = ByteUtils.Concat(BALANCE,from);
        byte[] from_balance = ctx.getObject(from_key);
        if (ByteUtils.Less(from_balance,token)){
            return Response.error("balance not enough");
        }
        byte[]allowance_key = ByteUtils.concat(ALLOWANCE,from,"_",to);
        byte[]allowance_balance = ByteUtils.add(ctx.getObject(allowance_key),token);
        ctx.putObject(allowance_key,allowance_balance);
        return Response.ok("approve succeed".getBytes());

    }

    public static void main(String[] args) {
        Driver.serve(new AwardManage());
    }
}